### DISCUSSION TITLE
Refetching of dependent queries

### DISCUSSION THREAD

**Discussion by JakubGIT** [TanStack/query#8027] [Q&A] [answered]
There [was already](https://github.com/TanStack/query/discussions/2612) a discussion about that, but unfortunately the solution doesn't work for me.

### The thing is:
- query1 is fetching and providing some data
- query 2 uses data from query 1 to fetch another piece of data
### What I want to have
- query 2 refetches if the data from query1 have changed
### Sandbox
- [link to sandbox](https://codesandbox.io/p/sandbox/53hw62?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522cm0p3iwcc00062a65xc4sb10f%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522cm0p3iwcc00022a65cak4s8fk%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522cm0p3iwcc00032a65jt87inyh%2522%257D%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522cm0p3iwcc00052a657mdl0ywj%2522%257D%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cm0p3iwcc00022a65cak4s8fk%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cm0p3iwcc00012a65ja6y0kfw%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Findex.tsx%2522%257D%255D%252C%2522id%2522%253A%2522cm0p3iwcc00022a65cak4s8fk%2522%252C%2522activeTabId%2522%253A%2522cm0p3iwcc00012a65ja6y0kfw%2522%257D%252C%2522cm0p3iwcc00052a657mdl0ywj%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cm0p3iwcc00042a65fr7o4alt%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%257D%255D%252C%2522id%2522%253A%2522cm0p3iwcc00052a657mdl0ywj%2522%252C%2522activeTabId%2522%253A%2522cm0p3iwcc00042a65fr7o4alt%2522%257D%252C%2522cm0p3iwcc00032a65jt87inyh%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522cm0p3iwcc00032a65jt87inyh%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)
- here you can find a working example using `useEffect()` to achieve that
  ```typescript
  const result1 = useQuery({
    queryFn: async () => {
      await new Promise((r) => setTimeout(r, 500));
      return data1;
    },
    queryKey: ["query", "1"],
  });

  const key = result1.data;

  const result2 = useQuery({
    queryFn: () => data2[key],
    queryKey: ["query", "2"],
    enabled: !!key,
  });

  useEffect(() => {
    result2.refetch();
  }, [result1.data]);
  ```
- when I change the source data and refetch query 1, the query 2 refetcher too and the value displayed changes
- is there any way to achieve this without `useEffect()`? When I comment out the `useEffect()`, the only way to get the new data into the DOM is to refetch query 1 and then to refetch query 2
- why do I have no luck when invalidating or refetching all queries? The buttons triggering refetching practically don't change anything

Many thanks

**Comment by TkDodo** [comment, marked as answer]
> query 2 refetches if the data from query1 have changed

then data from query1 needs to be part of the queryKey of query2. We have a lint rule that enforces this, I guess you aren't using it?

**Reply by JakubGIT** [reply]
Thanks a lot and sorry for the late reply. Yes, that works. 

I tried to integrate the linter once and somehow it didn't work - or there was nothing to lint, but I guess I didn't setup the eslint file correctly. I'll give it a try once more again.

**Reply by JustinGrote** [reply]
Sorry to necro but just wanted to note the lint rule helped me a lot. Question: If I do this, can I remove the `enabled: dependentQuery.isSuccess` from my dependent query? I'm wondering on initial load if it will see the data isn't populated and will wait or not.

**Reply by JustinGrote** [reply]
Also this makes the devtools a huge mess, if I pick like one dependent key is that enough?
<img width="237" height="256" alt="image" src="https://github.com/user-attachments/assets/bf63e0db-a5a2-432a-b18d-21cde780dca6" />


EDIT: Since the query is styleable I applied some amateur style to fix the wrap problem. Probably a better way to do this but it works good enough.
```
.tsqd-queries-overflow-container {
	overflow-x: auto;
	white-space: nowrap;
	flex: content;
	max-width: 70vw
}
```
